### PR TITLE
Added basic logging framework to handle args.quiet properly

### PR DESCRIPTION
**Pull request by SensibleSalmon**
(Please see #1024 for discussions of logging frameworks.)

Four levels: info, warn, err, debug

```
info -> stderr, quiet
err -> stderr, loud
debug -> stdout, quiet
warn -> stdout, loud
```

Open to suggestions on what does what.

Ref #790 and #1024

### CODE REVIEW

**Comment by bocajnotnef** [comment]
retest this, please

**Comment by ctb** [comment]
Please make the functions module level, and named `error`, `info`, `warning`, and `debug` (well, ok, the names can vary).

So, you would do:

    from khmer.logs import error, info, warning, debug

    ...

    error('File does not exist; exiting.)

In particular, no objects need to be instantiated, passed around, or otherwise harmed in this utopian vision of a stupidly lightweight logging interface.

**Comment by bocajnotnef** [comment]
So, how would I dynamically set verbosity? Would I just have a variable in the module that's "quiet" and I set it when we fire up a script?

**Comment by ctb** [comment]
On Mon, Jul 27, 2015 at 01:39:17PM -0700, Jake Fenton wrote:
> So, how would I dynamically set verbosity? Would I just have a variable in the module that's "quiet" and I set it when we fire up a script?

You have my permission to add a `configure_logging` function.

**Comment by ctb** [comment]
On Mon, Jul 27, 2015 at 11:40:53AM -0700, Jake Fenton wrote:
> Four levels: info, warn, err, debug
> 
>     Info is for status output, written to stderr, is surpressable
>     Warn is for warning output, written to stderr, is surpressable
> 
>     Err is for critical failures, will not be surpressed when quiet is set; Err
>     output is written to stderr
> 
>     Debug output is written to stdout and is never surpressed
> 
> Seems kind of silly to have two things that're the same (info and warn)--Open to suggestions on what does what.
> 
> Ref #790 and #1024 

'warn' gets printed out with '-q', I'd say.

**Comment by bocajnotnef** [comment]
I ended up changing what goes where; this is current: 

quiet == shushed on -q
loud == ignores -q
```
info -> stderr, quiet
err -> stderr, loud
debug -> stdout, quiet
warn -> stdout, loud
```

**Comment by bocajnotnef** [comment]
- [ ] Is it mergeable?
- [ ] Did it pass the tests?
- [ ] If it introduces new functionality in scripts/ is it tested?
  Check for code coverage with `make clean diff-cover`
- [ ] Is it well formatted? Look at `make pep8`, `make diff_pylint_report`,
  `make cppcheck`, and `make doc` output. Use `make format` and manual
  fixing as needed.
- [ ] Did it change the command-line interface? Only additions are allowed
  without a major version increment. Changing file formats also requires a
  major version number increment.
- [ ] Is it documented in the ChangeLog?
  http://en.wikipedia.org/wiki/Changelog#Format
- [ ] Was a spellchecker run on the source code and documentation after
  changes were made?
- [ ] Is the Copyright year up to date?

**Comment by bocajnotnef** [comment]
@ctb Shall I proceed with adding more scripts in, or get this merged?

**Comment by ctb** [comment]
get this reviewed :)

**Comment by bocajnotnef** [comment]
And on that note,

@ctb @luizirber @mr-c CR, please?

File: ChangeLog [hunk 1]
```diff
@@ -1,3 +1,8 @@
+2015-07-27  Jacob Fenton  <bocajnotnef@gmail.com>
+
+   * scrips/normalize-by-median.py,khmer/khmer_logger.py: added logging
+   framework, prorotyped in normalize-by-median
```
Comment (ctb) [inline ChangeLog]: prototyped

File: khmer/khmer_logger.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    if not __QUIET__:
+        print(message.format(**kwargs), file=sys.stderr)
+
+
+def log_err(message, **kwargs):
+    """For critical error output to stderr."""
+    print(message.format(**kwargs), file=sys.stderr)
+
+
+def log_dbg(message, **kwagrs):
```
Comment (ctb) [inline khmer/khmer_logger.py]: dbg => debug

File: khmer/khmer_logger.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    """For non-critical debug output to stderr."""
+    global __QUIET__
+    if not __QUIET__:
+        print(message.format(**kwargs), file=sys.stdout)
+
+
+def log_wrn(message, **kwargs):
```
Comment (ctb) [inline khmer/khmer_logger.py]: wrn => warn

File: khmer/khmer_logger.py [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+global __QUIET__
+__QUIET__ = False
+
+
+def configure_logging(quiet):
+    global __QUIET__
+    __QUIET__ = quiet
+
+
+def log_inf(message, **kwargs):
```
Comment (ctb) [inline khmer/khmer_logger.py]: inf => info

File: khmer/khmer_logger.py [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+    """For non-critical informative/status output to stderr."""
+    global __QUIET__
+    if not __QUIET__:
+        print(message.format(**kwargs), file=sys.stderr)
+
+
+def log_err(message, **kwargs):
```
Comment (ctb) [inline khmer/khmer_logger.py]: err => error

File: khmer/khmer_logger.py (hunk 4 above)
Comment (ctb) [inline khmer/khmer_logger.py]: also:

```
if kwargs:
   message = message.format(**kwargs)
```

Comment (bocajnotnef) [inline khmer/khmer_logger.py]: isn't necessary--works without the if.

Comment (ctb) [inline khmer/khmer_logger.py]: On Mon, Jul 27, 2015 at 03:26:27PM -0700, Jake Fenton wrote:
> > +# Contact: khmer-project@idyll.org
> > +#
> > +# Lightweight logging framework for khmer
> > +
> > +import sys
> > +
> > +global __QUIET__
> > +__QUIET__ = False
> > +
> > +
> > +def configure_logging(quiet):
> > +    global __QUIET__
> > +    __QUIET__ = quiet
> > +
> > +
> > +def log_inf(message, **kwargs):
> 
> isn't necessary--works without the if.

What happens with log_inf("Hello, {world}!")?

Comment (bocajnotnef) [inline khmer/khmer_logger.py]: Ahh. That borks.

Somehow I thought it borked without the .format().

**Comment by kdmurray91** [comment]
@bocajnotnef I'm sure you're cognisant of this, but please make sure that whenever stdout is used to output data, logging to stdout is automagically suppressed or redirected to stderr.

**Comment by bocajnotnef** [comment]
@ctb pointed out that it was a bad idea to have any logging related stuff
written to stdout. I was under the impression that there was existing info
being written to stdout that we'd want verbosity control over.

**Comment by camillescott** [comment]
Hmm. Why not just use Python's built in logging module? It's very simple
and takes care of things like global configuration.

https://docs.python.org/2/howto/logging.html#logging-basic-tutorial

**Comment by bocajnotnef** [comment]
Vetoed by @ctb--the quote used the phrase "festering pile of", as I recall.

**Comment by ctb** [comment]
@camillescott et al., please feel free to discuss logging frameworks here: #1024 . This PR is intended to  fix #790 without necessarily adding a big chunk of code or a new library dependency.

**Comment by ctb** [comment]
@bocajnotnef looks ok to me on a first pass. Main thought - do you want to make use of the implicit formatting of the log functions (e.g. line 88 in normalize-by-median), or keep the current mix?

**Comment by bocajnotnef** [comment]
Whoops, I meant to get all of those out. I think it's better if we stick with one format--personally, I think using the implicit formatting is cleaner.

**Comment by bocajnotnef** [comment]
@ctb Updated, ready for merge

**Comment by bocajnotnef** [comment]
retest this please

**Comment by bocajnotnef** [comment]
retest this please

**Comment by bocajnotnef** [comment]
retest this please

**Comment by bocajnotnef** [comment]
retest this please

**Comment by bocajnotnef** [comment]
retest this please

File: khmer/khmer_logger.py [hunk 6]
```diff
@@ -0,0 +1,53 @@
+from __future__ import print_function, unicode_literals
```
Comment (ctb) [inline khmer/khmer_logger.py]: can you move this down below the top comment, pls?

**Comment by bocajnotnef** [comment]
retest this please

**Comment by bocajnotnef** [comment]
Contrary to what conversation view says I did move that import.

@ctb @mr-c Merge, please?

**Comment by ctb** [comment]
Does `normalize-by-median.py -q data/100k-filtered.fa` (or any data file) work the way you expect?

File: scripts/normalize-by-median.py [hunk 7]
```diff
@@ -85,10 +86,10 @@ def __call__(self, reader, ifilename):
...
                     perc_kept = kept / float(total)
 
-                    print('... kept {kept} of {tot} or {perc_kept:.1%} so far'
-                          .format(kept=kept, tot=total, perc_kept=perc_kept),
-                          file=sys.stderr)
-                    print('... in file ' + ifilename, file=sys.stderr)
+                    log_info('... kept {kept} of {tot} or {perc_kept:.1%} so'
+                             'far', kept=kept, tot=total,
+                             perc_kept=perc_kept)
+                    log_info('... in file ' + ifilename)
```
Comment (ctb) [inline scripts/normalize-by-median.py]: can you switch this over to {} too?

Comment (ctb) [inline scripts/normalize-by-median.py]: (and elsewhere)

Comment (ctb) [inline scripts/normalize-by-median.py]: (lots of elsewhere)

**Comment by ctb** [comment]
Note: per conversation with @mr-c, I am happy to try to merge this for 2.0, but we will explore adding -q to more scripts/adding logging framework stuff/etc later, after we have some experience with +/- of this particular approach.

**Comment by ctb** [comment]
@bocajnotnef note: https://github.com/dib-lab/khmer/issues/1179#issuecomment-122671300 - I fixed this in 24c5397, please use that language in future.

**Comment by bocajnotnef** [comment]
@ctb WIlco re: language

+'s removed, implicit format in use; Merge?

**Comment by ctb** [comment]
I repeat: https://github.com/dib-lab/khmer/pull/1200#issuecomment-126343769

**Comment by bocajnotnef** [comment]
@ctb Sorry, missed that--fixed now. Merge?

**Comment by ctb** [comment]
Threw in some fixes; @mr-c @luizirber could you review, please?  The logging functionality is fine, be good to get a double-check on the tst utils change.

Also, @bocajnotnef, no more adding global khmer arguments if they're not being implemented throughout the codebase - this is strike 2 (see #1179 for strike 1).  I removed args.quiet because it was only used in normalize-by-median.py!

**Comment by bocajnotnef** [comment]
My understanding was that we were prototyping the use of --unique-kmers pre
#1179 and that the plan was to expand it to all scripts that interacted
with hash tables.

I had a similar understanding here--while Normalize-by-median is the
prototype we would later expand the functionality to all other scripts.

Additionally, the --quiet arg was in place in the hashbits args prior to my
meddling--I didn't add it.

That said, I should have pulled it down to just normalize's parser in this
PR since normalize is the only one with the framework to handle it, but it
appeared to silence some output, just not all the things.

**Comment by ctb** [comment]
On Sat, Aug 01, 2015 at 09:00:24AM -0700, Jake Fenton wrote:
> My understanding was that we were prototyping the use of --unique-kmers pre
> #1179 and that the plan was to expand it to all scripts that interacted
> with hash tables.
> 
> I had a similar understanding here--while Normalize-by-median is the
> prototype we would later expand the functionality to all other scripts.

Yep, that's the plan, but (for both this and the --unique-kmers problem)
each PR should be internally consistent.  Globally adding arguments that do
nothing to most scripts is bad...

> Additionally, the --quiet arg was in place in the hashbits args prior to my
> meddling--I didn't add it.

Huh, ok, apologies :)

> That said, I should have pulled it down to just normalize's parser in this
> PR since normalize is the only one with the framework to handle it, but it
> appeared to silence some output, just not all the things.

Yep.

Doesn't seem like I broke any tests so *shrug*.

**Comment by bocajnotnef** [comment]
Yeah. Herein I'll make sure to contain things a bit more. Makes sense not
to have args strewn across the code that we don't react to.

**Comment by ctb** [comment]
Hah, I reverted the changes I made that needed review, so I think I'll merge (https://github.com/dib-lab/khmer/pull/1200#issuecomment-126926851).

**Comment by bocajnotnef** [comment]
Thanks!

File: khmer/khmer_logger.py [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+
+
+def log_error(message, **kwargs):
+    """For critical error output to stderr."""
+    if kwargs:
+        message = message.format(**kwargs)
+    print(message, file=sys.stderr)
+
+
+def log_debug(message, **kwagrs):
```
Comment (mr-c) [inline khmer/khmer_logger.py]: `kwargs`

Please write a test that exercises this method and check for test coverage. Also, please review the pylint report for this file, your `global` usage is off.

Comment (ctb) [inline khmer/khmer_logger.py]: > Please write a test that exercises this method and check for test coverage. Also, please review the pylint report for this file, your `global` usage is off.

I think file an issue, at this point.

**Pull request merged**